<a href="https://colab.research.google.com/github/MarioVidal-Co/Analitica/blob/main/Reto_2_Mario_Vidal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier

from google.colab import files
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
nxl='/content/drive/MyDrive/Análitica de Negocios/2. BD1_Travel_Data_int.xlsx'
XDB=pd.read_excel(nxl,sheet_name=1)
XDB=XDB.dropna()
XDB.head(100)

XD=np.array(XDB[['College_B','FoodSpend','Income','TravelSpend']])
print(XD)
yd=np.array(XDB[['Credit_Card']])
print(yd)

[[1.00000e+00 5.47243e+03 4.91500e+04 8.27400e+02]
 [0.00000e+00 9.13073e+03 4.78060e+04 8.63550e+02]
 [0.00000e+00 4.45067e+03 4.60500e+04 1.18005e+03]
 ...
 [0.00000e+00 3.58586e+03 4.22500e+04 2.25195e+03]
 [1.00000e+00 1.83395e+03 6.02480e+04 2.00025e+03]
 [0.00000e+00 6.20597e+03 5.85200e+04 3.66750e+03]]
[[1]
 [1]
 [1]
 [1]
 [1]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [1]
 [1]
 [1]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [1]
 [0]
 [1]
 [1]
 [0]
 [1]
 [0]
 [1]
 [1]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [1]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [1]
 [0]
 [1]
 [0]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,yd)
ydp=mar.predict(XD)

from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
print(cm)

VN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
VP=cm[1,1]

Ex=(VN+VP)/(VN+VP+FN+FP)
print("La Exactitud alcanzada por el modelo es:",Ex)
Se= (VP)/(FN+VP)
print("La sensibilidad alcanzada por el modelo es:",Se)
Es= (VN)/(VN+FP)
print("La Especificidad alcanzada por el modelo es:",Es)

[[334   9]
 [ 42 115]]
La Exactitud alcanzada por el modelo es: 0.898
La sensibilidad alcanzada por el modelo es: 0.732484076433121
La Especificidad alcanzada por el modelo es: 0.9737609329446064


In [ ]:
from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data

vs=['College_B','FoodSpend','Income','TravelSpend']
dot_graph=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_graph)
graph.write_png('arbol2.png')

True

In [ ]:
XDB2=pd.read_excel(nxl,sheet_name=1)
XDB2=XDB2.dropna()
XDB2.head(100)

XD2=np.array(XDB2[['College_B','FoodSpend','Income','TravelSpend']]) #Datos de entrada
print(XD2)

#Pronosticamos
ydp2=mar.predict(XD2)
print(ydp2)
print("El número total de personas que usarían su TC es:",np.sum(ydp2))
print("El número total de datos es:",len(ydp2))
print("El número de personas que usarían su TC es:",len(ydp2)-np.sum(ydp2))

[[1.00000e+00 5.47243e+03 4.91500e+04 8.27400e+02]
 [0.00000e+00 9.13073e+03 4.78060e+04 8.63550e+02]
 [0.00000e+00 4.45067e+03 4.60500e+04 1.18005e+03]
 ...
 [0.00000e+00 3.58586e+03 4.22500e+04 2.25195e+03]
 [1.00000e+00 1.83395e+03 6.02480e+04 2.00025e+03]
 [0.00000e+00 6.20597e+03 5.85200e+04 3.66750e+03]]
[1 1 1 1 1 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 1 0 1 0 0 1 0 1 0 0 1 0 1 1 0
 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 1 1 0 0 0 1 0 0 1 0 1 0 0 0 0 0 0 1
 0 0 1 0 0 0 0 0 1 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 1 0 0
 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0
 0 1 1 0 0 1 1 1 0 0 0 1 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 1 1 0 0 0 1 0 0 0
 1 0 0 0 0 0 0 1 0 0 0 1 0 0 0 1 1 0 0 1 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0
 0 0 1 1 0 0 0 0 0 0 0 0 1 0 0 1 0 0 1 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 1 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0
 0 0 0 0 0 0 0 0 1 1 0 1 1 1 0 0 1 1 0 1 0 0 0 1 0 0 0 0 1 1 0 0 0 1 0 1 0
 0 0 0 0 0 0 

**Análisis de Resultados**

De acuerdo con la base de datos suministrada y despues de crear el modelo de decisión, podemos concluir que:

Este modelo presenta una exactitud del 89.8%, lo cual nos permite determinar con menor probabilidad de error si una persona usará o no su Tarjeta de Crédito durante un viaje. Con respecto a la sensibilidad, este modelo nos permite alcanzar un valor de 73,24%, lo cual permite categorizar correctamente si una persona utiliza su tarjeta de crédito. De igual forma la especificidad de este modelo alcanzó un 97.37%, indicando que este modelo permite clasificar correctamente a aquellas personas que no usarán su Tarjeta de crédito durante un viaje.

Se destacan tres Nodos Puros con las siguientes carácterísticas:

* [1,0] (SI TravelSpend<=1188 AND FoodSpend>4972.55)
* [0,47] (SI TravelSpend<=1188 AND FoodSpend>4990.73)
* [108,0] (SI TravelSpend>1567.875 AND Income>59098)

Con respecto a los nuevos clientes, se pueden apreciar los siguientes casos:

* *Primer Cliente:* SI TravelSpend>1567.875 AND Income 59098, la cual nos deja con un nodo puro de [108,0] lo que indica que este cliente tiene una posibilidad del 100% de no usar su tarjeta de crédito durante su viaje.
* *Segundo Cliente:* SI Income < 59098 AND FoodSpend > 2711.895 AND TravelSpend < 3008.25, esta persona tiene una posibilidad del 78.67% de no usar su tarjeta de crédito durante su viaje.